src/background/extension.js

```diff
@@ -94,42 +57,105 @@ export class Extension {
       state.activateTab(/** @type {number} */ (tab.id));
     };
 
-    async function restoreSavedTabState() {
+    /**
+     * Initialize cached state for browser tabs by querying each tab to see
+     * whether the extension is active there.
+     *
+     * This should be called when the extension is loaded or reloaded.
+     */
+    const initTabStates = async () => {
       const tabs = await chromeAPI.tabs.query({});
-      const tabIds = tabs
-        .filter(tab => tab.id !== undefined)
-        .map(({ id }) => /** @type {number} */ (id));
-      store.reload(tabIds);
-      state.load(store.all());
-      tabIds.forEach(tabId => {
-        onTabStateChange(tabId, state.getState(tabId));
-      });
-    }
+      const activeStates = await Promise.all(
+        tabs.map(async tab => {
+          if (tab.id === undefined) {
+            return false;
+          }
+          try {
+            const active = await sidebar.isClientActiveInTab(tab);
+            return active;
+          } catch (e) {
+            console.warn(
+              `Unable to determine extension state in tab ${tab.id}`,
+              e
+            );
+            return false;
+          }
+        })
+      );
+
+      for (let i = 0; i < tabs.length; i++) {
+        const tab = tabs[i];
+        if (!tab.id) {
+          continue;
+        }
+        const isActive = activeStates[i];
+
+        // nb. If tab status is not available, we optimistically assume it is
+        // loaded.
+        const ready =
+          tab.status === 'complete' || typeof tab.status !== 'string';
+
+        state.setState(tab.id, {
+          state: isActive ? 'active' : 'inactive',
+          extensionSidebarInstalled: isActive,
+          ready,
+        });
+      }
+    };
+
+    /**
+     * Initialize the extension.
+     *
+     * This queries the state of the extension in existing tabs and sets up
+     * event listeners to respond to future tab changes.
+     *
+     * If the extension's state in a particular tab cannot be determined,
+     * the extension is assumed not to be loaded in that tab.
+     *
+     * @return - A promise that resolves once listeners have been set up and
+     *   the state of existing tabs has been determined.
+     */
+    this.init = async () => {
```

Comment:
Maybe this could be defined as a regular class method instead of a callable property?

Reply:
In this PR I stuck to the existing style of this module, but in future this code would benefit from some modernization. You'll see this elsewhere in the browser-extension codebase too, places where pre-ES6-isms still exist.